Corona-Warn-App/src/main/java/de/rki/coronawarnapp/playbook/Playbook.kt

```diff
@@ -19,23 +32,210 @@ import de.rki.coronawarnapp.server.protocols.internal.pt.CheckInOuterClass
+                    cause = exception
+                )
+            )
+        }
+    }
+
+    /**
+     * Distinguish BadRequestException to present more insightful message to the end user
+     */
+    private fun wrapException(exception: Exception?) = when (exception) {
+        is BadRequestException -> TanPairingException(
+            code = exception.statusCode,
+            message = "Tan has been retrieved before for this registration token",
+            cause = exception
+        )
+        else -> exception
+    }
+
+    private suspend fun dummy(launchFollowUp: Boolean) {
+        // fake verification
+        ignoreExceptions { verificationServer.retrieveTanFake() }
+
+        // fake verification
+        ignoreExceptions { verificationServer.retrieveTanFake() }
+
+        // fake submission
+        ignoreExceptions { submissionServer.submitFakePayload() }
+
+        if (launchFollowUp)
+            coroutineScope.launch { followUpPlaybooks() }
+    }
+
+    suspend fun dummy() = dummy(true)
+
+    private suspend fun followUpPlaybooks() {
+        val runsToExecute = IntRange(
+            MIN_NUMBER_OF_SEQUENTIAL_PLAYBOOKS - 1 /* one was already executed */,
+            MAX_NUMBER_OF_SEQUENTIAL_PLAYBOOKS - 1 /* one was already executed */
+        ).random()
+        Timber.i("[$uid] Follow Up: launching $runsToExecute follow up playbooks")
+
+        repeat(runsToExecute) {
+            val executionDelay = IntRange(
+                MIN_DELAY_BETWEEN_SEQUENTIAL_PLAYBOOKS,
+                MAX_DELAY_BETWEEN_SEQUENTIAL_PLAYBOOKS
+            ).random()
+            Timber.i("[$uid] Follow Up: (${it + 1}/$runsToExecute) waiting $executionDelay[s]...")
+            delay(TimeUnit.SECONDS.toMillis(executionDelay.toLong()))
+
+            dummy(false)
+        }
+        Timber.i("[$uid] Follow Up: finished")
+    }
+
+    private suspend fun ignoreExceptions(body: suspend () -> Unit) {
+        try {
+            body.invoke()
+        } catch (e: Exception) {
+            Timber.v("Ignoring dummy request exception: %s", e.toString())
+        }
+    }
+
+    private suspend fun <T> executeCapturingExceptions(body: suspend () -> T): Pair<T?, Exception?> {
+        return try {
+            val result = body.invoke()
+            result to null
+        } catch (e: Exception) {
+            null to e
+        }
+    }
+
+    private fun propagateException(vararg exceptions: Exception?): Nothing {
+        throw exceptions.filterNotNull().firstOrNull() ?: IllegalStateException()
+    }
+
+    companion object {
+        const val MIN_NUMBER_OF_SEQUENTIAL_PLAYBOOKS = 1
+        const val MAX_NUMBER_OF_SEQUENTIAL_PLAYBOOKS = 1
+        const val MIN_DELAY_BETWEEN_SEQUENTIAL_PLAYBOOKS = 0
+        const val MAX_DELAY_BETWEEN_SEQUENTIAL_PLAYBOOKS = 0
+    }
 
     data class SubmissionData(
         val registrationToken: String,
+        val authCode: String? = null,
```

Comment:
I think this should not be nullable , it is required for submission otherwise we don't submit

Reply:
It is null until we were successful with exchanging it for the registration token, so it has to be nullable

Reply:
Ah, you are right, I confused with the test data, will adapt it